lib/core/src/Cardano/Pool/Rank.hs

```diff
@@ -0,0 +1,379 @@
+the changes to its cost, margin and pledge will not be visible until the
+next epoch; put differently, the rewards for the stake snapshot taken
+at the end of epoch /e1/ will only depend on changes to cost, margin,
+and pledge that the pool owner initiated in epoch /e0/.
+This prevents pool owners from duping delegators by changing pool costs
+during an epoch. However, the pool owner could still choose to undelegate
+his stake, and fail to meet his pledge at the end of epoch /e1/,
+which results in zero rewards paid out at the beginning of /e5/.
+
+To summarize, in order to make an informed delegation choice
+during epoch /e1/, the delegator needs to know 'RewardInfoPool' where
+
+* 'stakeRelative', 'ownerStake', and 'ownerStakeRelative' are
+   taken at the time of decision (in epoch /e1/).
+* 'ownerPledge', 'margin', 'cost' are the values of the last pool
+   registration certificate
+   from epoch /e0/ in case of an update,
+   or from epoch /e1/ in case of a newly created pool.
+
+For the 'performanceEstimate', it's best to estimate it from recent pool
+block production using functions provided here.
+
+-}
+
+-- | Summary of stake distribution and stake pools obtained from network.
+data StakePoolsSummary = StakePoolsSummary
+    { rewardParams :: RewardParams
+    , pools :: Map PoolId RewardInfoPool
+    } deriving (Show, Eq)
+
+instance Buildable StakePoolsSummary where
+    build StakePoolsSummary{rewardParams,pools} = blockListF' "" id
+        [ "Global reward parameters: " <> build rewardParams
+        , "Individual pools: " <> mapF (Map.toList pools)
+        ]
+
+{-------------------------------------------------------------------------------
+    Reward formulas
+-------------------------------------------------------------------------------}
+fractionOf :: RealFrac r => r -> Coin -> Coin
+fractionOf r (Coin x) = Coin . floor $ r * fromIntegral x
+
+proportionTo :: Coin -> Coin -> Rational
+proportionTo _        (Coin 0) = 0
+proportionTo (Coin x) (Coin y) = fromIntegral x / fromIntegral y
+
+z0 :: RewardParams -> Rational
+z0 RewardParams{nOpt} = 1 / fromIntegral nOpt
+
+epochsPerYear :: Int
+epochsPerYear = 73
+
+-- | The current (yearly) return per unit of stake.
+currentROS :: RewardParams -> RewardInfoPool -> Coin -> Percentage
+currentROS rp RewardInfoPool{..} x
+    | ownerStake < ownerPledge = clipToPercentage 0
+    | otherwise = clipToPercentage $ (1 + astar)^epochsPerYear - 1
+  where
+    s     = clipToPercentage $ ownerPledge `proportionTo` (totalStake rp)
+    sigma = getPercentage stakeRelative + (x `proportionTo` totalStake rp)
+
+    astar
+        | sigma == 0 = 0
+        | otherwise  = shareAfterFees (1/sigma) cost margin fhat
+            `proportionTo` totalStake rp
+    fhat  = optimalRewards rp s sigma
+
+-- | The (yearly) return per unit of stake
+-- for a pool that has reached saturation.
+saturationROS :: RewardParams -> RewardInfoPool -> Percentage
+saturationROS rp RewardInfoPool{..}
```

Comment:
saturationROS and currentROS are the same, except for sigma, right?

Reply:
Yes.